webcrypto/subtle_crypto.go

```diff
@@ -35,9 +36,71 @@ type SubtleCrypto struct {
 // The `key` parameter should be a `CryptoKey` to be used for encryption.
 //
 // The `data` parameter should contain the data to be encryption.
-func (sc *SubtleCrypto) Encrypt(algorithm goja.Value, key goja.Value, data []byte) *goja.Promise {
-	// TODO: implementation
-	return nil
+func (sc *SubtleCrypto) Encrypt(algorithm, key, data goja.Value) *goja.Promise {
+	rt := sc.vu.Runtime()
+	promise, resolve, reject := sc.makeHandledPromise()
+
+	// 2.
+	// We obtain a copy of the key data, because we might need to modify it.
+	plaintext, err := exportArrayBuffer(rt, data)
+	if err != nil {
+		reject(err)
+		return promise
+	}
+
+	// 3.
+	normalized, err := normalizeAlgorithm(rt, algorithm, OperationIdentifierEncrypt)
+	if err != nil {
+		reject(err)
+		return promise
+	}
+
+	var ck CryptoKey
+	if err = rt.ExportTo(key, &ck); err != nil {
+		reject(NewError(0, TypeError, "key argument does hold not a valid CryptoKey object"))
+		return promise
+	}
+
+	// 8.
+	keyAlgorithmName := key.ToObject(rt).Get("algorithm").ToObject(rt).Get("name").String()
+	if normalized.Name != keyAlgorithmName {
+		reject(NewError(0, InvalidAccessError, "algorithm name does not match key algorithm name"))
+		return promise
+	}
+
+	encrypter, err := newEncryptDecrypter(rt, normalized, algorithm)
+	if err != nil {
+		reject(err)
+		return promise
+	}
+
+	go func() {
+		// 9.
+		for !containsUsage(ck.Usages, EncryptCryptoKeyUsage) {
+			reject(NewError(0, InvalidAccessError, "key does not contain the 'encrypt' usage"))
+			return
+		}
+
+		var ciphertext []byte
+
+		switch normalized.Name {
+		case AESCbc, AESCtr, AESGcm:
+			// 10.
+			ciphertext, err = encrypter.Encrypt(plaintext, ck)
+			if err != nil {
+				reject(err)
```

Comment:
Is an error generated here, clear enough or should we wrap it?

Reply:
I have been careful to do so indeed, and double checked, in every cases, the Encrypter will return a rich error 👍🏻 The specification does not expect the operation to fail, so we have free hands there. I'd prefer to keep this granularity as it would help debugging user issues if they occur. Feel free to unresolve if you would prefer it another way 🙇🏻